source/geod24/concurrency.d

```diff
@@ -985,6 +986,161 @@ protected:
         bool notified;
     }
 
+    class Channel (T)
+    {
+        this(ulong size = 0)
+        {
+            this.size = size;
+            m_lock = new Mutex;
+        }
+
+        bool write(T val)
+        {
+            bool notify_peer, fiber_was_blocked = false;
+            ChannelQueueEntry qEnt;
+
+            synchronized (m_lock)
+            {
+                notify_peer = m_rxq.length > 0;
+
+                if (m_closed)
+                {
+                    return false;
+                }
+                else if (m_rxq.length > 0)
+                {
+                    qEnt = m_rxq[0];
+                    m_rxq = m_rxq.remove(0);
+
+                    *(qEnt.pVal) = val;
+                }
+                else if (m_buffer.length < size)
+                {
+                    m_buffer ~= val;
+                }
+                else
+                {
+                    qEnt = new ChannelQueueEntry(&val);
+                    m_txq ~= qEnt;
+                }
+            }
+
+            if (notify_peer)
+            {
+                qEnt.sem.notify();
+            }
+            else if (qEnt)
+            {
+                fiber_was_blocked = true;
+                qEnt.sem.wait();
+            }
+
+            return !(fiber_was_blocked) || qEnt.success;
```

Comment:
```suggestion
                qEnt.sem.wait();
                return false;  // fiber was blocked
            }

            returnq Ent.success;
```

It removes the need to have an additional state variable.

Reply:
Getting blocked doesn't always mean a failure. This doesn't cover the case where the caller gets blocked and eventually woken up by a read/write. In that case we would need to return true.